### ISSUE TITLE
Problems during compiling derecho demo apps

### ISSUE THREAD

**Issue by Steamgjk** [Derecho-Project/derecho#230]
Hi, Derecho staff.
I am studying Derecho and I followed the instructions and I can successsfully compile drecho and install it to my VM.

Then I continued to compile the app in , but ecnounted a problem.
I am just using some typical command:

cd demo
mkdir build
cd build
cmake ..
make

<img width="641" alt="图片" src="https://user-images.githubusercontent.com/7100589/159174632-afe56a68-d3c6-4cc9-9812-babe3fc33ac7.png">

<img width="875" alt="图片" src="https://user-images.githubusercontent.com/7100589/159174686-377e5ad8-5fd8-4b9b-86f8-1cf18ea2799c.png">
<img width="869" alt="图片" src="https://user-images.githubusercontent.com/7100589/159174707-83acff0c-373e-4441-87e4-d40439aa9e8d.png">

I attach the whole error logs

[error-log-derecho.txt](https://github.com/Derecho-Project/derecho/files/8311820/error-log-derecho.txt)

Then I turn to use the method from the Readme

g++ -std=c++1z -o simple_replicated_objects  simple_replicated_objects.cpp -lderecho -pthread -lcrypto

(After adding -lcrypto, I can successfully compile the application simple_replicated_objects, so I think the failure of cmake method is also related to missing some dependent libs) 
I can continue to study the application with g++. But would you please check why cmake method cannot work? (I think cmake is more convenient than g++, because I donot need to compile apps one by one with cmake)

**Comment by KenBirman** [comment]
The thing that jumps out to me is that you used std=C++1z and then got a g++ complaint about features defined only in C++17 and later.  My speculation would be that C++1z somehow isn't a true, complete, C++17.  Try again but tell it to actually use C++17.  I'm saying this because (1) Derecho requires C++17 or later, and (2) from the error message, it obviously doesn't think it was told to use C++17.

When working from source and compiling using your own commands, you may run into issues such as this, tied to how you've configured your toolchain (meaning the compiler standard, the linker, etc).  All of them need to be C++17 or later.

**Comment by Steamgjk** [comment]
> The thing that jumps out to me is that you used std=C++1z and then got a g++ complaint about features defined only in C++17 and later. My speculation would be that C++1z somehow isn't a true, complete, C++17. Try again but tell it to actually use C++17. I'm saying this because (1) Derecho requires C++17 or later, and (2) from the error message, it obviously doesn't think it was told to use C++17.
> 
> When working from source and compiling using your own commands, you may run into issues such as this, tied to how you've configured your toolchain (meaning the compiler standard, the linker, etc). All of them need to be C++17 or later.

Hi, @KenBirman 
I am afraid that is not the reason.
<img width="878" alt="图片" src="https://user-images.githubusercontent.com/7100589/159177779-db17bec3-e83e-420e-aab8-0763cdb58516.png">

Even if I change from  -std=c++1z to  -std=c++17, if I do not add -lcrypto, it will report the same error.
Besides, I am just using a clean VM from Google Cloud with Ubuntu 20.04. I just followed the derecho ReadMe to install the dependencies. After that, I did not have any other configurations ( the compiler standard, the linker, etc). I think it satisfies "All of them need to be C++17 or later".

But that is not a big issue. So long as I add -lcrypto in the g++ command, I can successfully compile the app. 

Regarding the issue with cmake, I think I found the reasons. I have configured cmake -DCMAKE_BUILD_TYPE=Release,
so I can only make applications in the Release folder. I can successfully built the apps now. 
If I create a different folder and make there, maybe the path configuration will be wrong,  causing some missing deps. 

<img width="857" alt="图片" src="https://user-images.githubusercontent.com/7100589/159178102-60a7bfc0-c46f-4f3a-8d2e-e0f0bac38bcd.png">

<img width="712" alt="图片" src="https://user-images.githubusercontent.com/7100589/159178250-62defaa4-c0de-43cf-b748-1898f54b95f3.png">


By the way, in the ReadMe, you staff suggest using -std=c++1z. 

<img width="585" alt="图片" src="https://user-images.githubusercontent.com/7100589/159177859-e441e326-db58-4a2f-b73f-c10a848dc0f1.png">

I feel there is no much difference between c++17 and c++1z in practice, any special consideration about that? Which one is recommended?
